File: frontend/src/types/backtest.ts
// ─────────────────────────────────────────────────────────────────────────
// Form state types
// ─────────────────────────────────────────────────────────────────────────

export interface RangeParam {
  min: number;
  max: number;
  step: number;
}

/** Cross-product factor (used in mode='multi'). */
export interface FactorConfig {
  // ── factor-only ──
  indicator: string;
  strategy: string;
  data_column: string;
  window_range: RangeParam;
  signal_range: RangeParam;
  symbol?: string;         // internal_cusip for cross-product factor (pair trade)
  vendor_symbol?: string;  // direct vendor symbol override for this factor
  data_source?: string;    // per-factor data source override
}

/** Top-level backtest form. */
export interface BacktestConfig {
  // ── trading product (product-only) ──
  symbol: string;       // internal_cusip from product dropdown
  vendorSymbol: string; // direct vendor symbol override (e.g. BTC-USD)
  dataSource: string;
  assetType: string;

  // ── shared controls (apply to product + every factor) ──
  start: string;
  end: string;
  tradingPeriod: number;
  feeBps: number;
  /** When true, refetch all product+factor data from the provider and
   *  insert a new BT.API_REQUEST version. When false (default), serve
   *  from cache only — backend returns 400 if the cache misses. */
  refreshDataset: boolean;

  // ── single-factor mode ──
  mode: 'single' | 'multi';
  indicator: string;
  strategy: string;
  windowRange: RangeParam;
  signalRange: RangeParam;

  // ── multi-factor mode ──
  conjunction: string;
  factors: FactorConfig[];

  // ── walk-forward (analysis option) ──
  walkForward: boolean;
  splitRatio: number;
}

// ─────────────────────────────────────────────────────────────────────────
// API request types (snake_case to match backend Pydantic models)
// ─────────────────────────────────────────────────────────────────────────

export interface OptimizeRequest {
  // ── trading product ──
  symbol: string;
  data_source?: string;

  // ── shared (product + factors) ──
  start: string;
  end: string;
  trading_period: number;
  fee_bps: number;
  refresh_dataset?: boolean;

  // ── strategy/mode ──
  mode: 'single' | 'multi';

  // ── single-factor ──
  indicator?: string;
  strategy?: string;
  window_range?: RangeParam;
  signal_range?: RangeParam;

  // ── multi-factor ──
  conjunction?: string;
  factors?: FactorConfig[];

  // ── walk-forward (run inline when true) ──
  walk_forward?: boolean;
  split_ratio?: number;
}

export interface PerformanceRequest {
  // ── trading product ──
  symbol: string;
  data_source?: string;

  // ── shared ──
  start: string;
  end: string;
  trading_period: number;
  fee_bps: number;
  refresh_dataset?: boolean;

  // ── strategy/mode ──
  mode: 'single' | 'multi';

  // ── single-factor ──
  indicator?: string;
  strategy?: string;
  window?: number;
  signal?: number;

  // ── multi-factor ──
  conjunction?: string;
  factors?: FactorConfig[];
  windows?: number[];
  signals?: number[];
}

/**
 * One row of optimization results. Discriminated by the presence of
 * `window`/`signal` (single-factor) vs `window_0`/`signal_0`/... (multi-factor).
 *
 * Use the helpers in `utils/top10.ts` (`isSingleFactorRow`, `multiFactorParams`)
 * to read window/signal values safely instead of `as number` casts.
 */
export type Top10Row = SingleFactorRow | MultiFactorRow;

export interface SingleFactorRow {
  window: number;
  signal: number;
  sharpe: number;
  /** Reserved for future per-row metadata; not used for window/signal lookups. */
  [key: string]: number | string | null | undefined;
}

export interface MultiFactorRow {
  sharpe: number;
  /**
   * Per-factor params live under `window_0`, `signal_0`, `window_1`, `signal_1`, ...
   * Keep the index signature wide enough to allow that, but narrow values to numbers.
   */
  [key: string]: number | string | null | undefined;
}

export interface OptimizeResponse {
  total_trials: number;
  valid: number;
  best: Top10Row;
  top10: Top10Row[];
  grid: Top10Row[];
  performance?: PerformanceResponse;
  walk_forward?: WalkForwardResponse;
}

export interface OptimizeProgress {
  trial: number;
  total: number;
  best_sharpe: number | null;
}

export interface EquityPoint {
  datetime: string;
  cumu: number;
  buy_hold_cumu: number;
  dd: number;
  buy_hold_dd: number;
}

export interface PerformanceResponse {
  strategy_metrics: Record<string, number>;
  buy_hold_metrics: Record<string, number>;
  equity_curve: EquityPoint[];
  perf_csv: string;
}

// Walk-forward analysis
export interface WalkForwardRequest extends OptimizeRequest {
  split_ratio: number;
}

export interface WalkForwardResponse {
  best_window: number | number[];
  best_signal: number | number[];
  is_metrics: Record<string, number>;
  oos_metrics: Record<string, number>;
  overfitting_ratio: number | null;
  equity_curve: EquityPoint[];
  split_date: string;
}

File: api/schemas/backtest.py
import math

import numpy as np
from pydantic import BaseModel


class RangeParam(BaseModel):
    min: float
    max: float
    step: float

    def to_values(self, as_int: bool = False) -> tuple:
        """Expand into a concrete value sequence for the optimizer."""
        if as_int:
            return tuple(range(int(self.min), int(self.max) + 1, int(self.step)))
        return tuple(np.arange(self.min, self.max + self.step / 2, self.step))


class FactorConfig(BaseModel):
    indicator: str
    strategy: str
    data_column: str = "price"
    window_range: RangeParam
    signal_range: RangeParam
    symbol: str | None = None          # internal_cusip for cross-product factor
    vendor_symbol: str | None = None   # direct vendor symbol override
    data_source: str | None = None     # per-factor data source override


# ── Requests ────────────────────────────────────────────────────────

class DataRequest(BaseModel):
    symbol: str
    start: str
    end: str


class OptimizeRequest(BaseModel):
    symbol: str
    start: str
    end: str
    mode: str  # "single" | "multi"
    trading_period: int
    fee_bps: float = 5.0
    data_source: str = "yahoo"  # REFDATA.APP.NAME
    # Single-factor
    indicator: str | None = None
    strategy: str | None = None
    window_range: RangeParam | None = None
    signal_range: RangeParam | None = None
    # Multi-factor
    conjunction: str | None = None
    factors: list[FactorConfig] | None = None
    # Walk-forward (run inline with optimization when True)
    walk_forward: bool = False
    split_ratio: float = 0.5


class PerformanceRequest(BaseModel):
    symbol: str
    start: str
    end: str
    mode: str
    trading_period: int
    fee_bps: float = 5.0
    data_source: str = "yahoo"  # REFDATA.APP.NAME
    # Single-factor
    indicator: str | None = None
    strategy: str | None = None
    window: int | None = None
    signal: float | None = None
    # Multi-factor
    conjunction: str | None = None
    factors: list[FactorConfig] | None = None
    windows: list[int] | None = None
    signals: list[float] | None = None


class WalkForwardRequest(OptimizeRequest):
    split_ratio: float = 0.5


# ── Responses ───────────────────────────────────────────────────────

class DataResponse(BaseModel):
    rows: int
    start_date: str
    end_date: str
    data: list[dict]


class EquityPoint(BaseModel):
    datetime: str
    cumu: float
    buy_hold_cumu: float
    dd: float
    buy_hold_dd: float


class PerformanceResponse(BaseModel):
    strategy_metrics: dict
    buy_hold_metrics: dict
    equity_curve: list[EquityPoint]
    perf_csv: str


class WalkForwardResponse(BaseModel):
    best_window: int | list[int]
    best_signal: float | list[float]
    is_metrics: dict
    oos_metrics: dict
    overfitting_ratio: float | None
    equity_curve: list[EquityPoint]
    split_date: str


class OptimizeResponse(BaseModel):
    total_trials: int
    valid: int
    best: dict
    top10: list[dict]
    grid: list[dict]
    optuna_plots: dict | None = None
    # Inline performance for best params (when available)
    performance: PerformanceResponse | None = None
    # Inline walk-forward (when walk_forward=True in request)
    walk_forward: WalkForwardResponse | None = None
